gateway/ln-gateway/src/lnd.rs

```diff
@@ -510,6 +513,147 @@ impl ILnRpcClient for GatewayLndClient {
+                    info!("LND payment request failed for invoice {invoice:?} with {status:?}");
+                    LightningRpcError::FailedPayment {
+                        failure_reason: format!("Failed to make outgoing payment {status:?}"),
+                    }
+                })?;
+
+            debug!(
+                "LND payment request sent for invoice {invoice:?}, waiting for payment status..."
+            );
+            let mut messages = payments.into_inner();
+            loop {
+                match messages
+                    .message()
+                    .await
+                    .map_err(|error| LightningRpcError::FailedPayment {
+                        failure_reason: format!("Failed to get payment status {error:?}"),
+                    }) {
+                    Ok(Some(payment)) if payment.status() == PaymentStatus::Succeeded => {
+                        info!("LND payment succeeded for invoice {invoice:?}");
+                        break bitcoin_hashes::hex::FromHex::from_hex(
+                            payment.payment_preimage.as_str(),
+                        )
+                        .map_err(|error| {
+                            LightningRpcError::FailedPayment {
+                                failure_reason: format!("Failed to convert preimage {error:?}"),
+                            }
+                        })?;
+                    }
+                    Ok(Some(payment)) if payment.status() == PaymentStatus::InFlight => {
+                        debug!("LND payment is inflight");
+                        continue;
+                    }
+                    Ok(Some(payment)) => {
+                        info!("LND payment failed for invoice {invoice:?} with {payment:?}");
+                        let failure_reason = payment.failure_reason();
+                        return Err(LightningRpcError::FailedPayment {
+                            failure_reason: format!("{failure_reason:?}"),
+                        });
+                    }
+                    Ok(None) => {
+                        info!("LND payment failed for invoice {invoice:?} with no payment status");
+                        return Err(LightningRpcError::FailedPayment {
+                            failure_reason: format!(
+                                "Failed to get payment status for payment hash {:?}",
+                                invoice.payment_hash
+                            ),
+                        });
+                    }
+                    Err(e) => {
+                        info!("LND payment failed for invoice {invoice:?} with {e:?}");
+                        return Err(e);
+                    }
+                }
+            }
+        };
+        Ok(PayInvoiceResponse { preimage })
+    }
+
+    /// Returns true if the lightning backend supports payments without full
+    /// invoices
+    fn supports_private_payments(&self) -> bool {
+        true
+    }
```

Comment:
@justinmoon We could make this configurable in case there is a problem with the private pay implementation.